src/btree/row_srch.c

```diff
@@ -614,14 +620,7 @@ leaf_match:	cbt->compare = 0;
 
 	return (0);
 
-err:	/*
-	 * Release the current page if the search started at the root. If the
-	 * search didn't start at the root we should never have gone looking
-	 * beyond the start page.
-	 */
-	WT_ASSERT(session, leaf == NULL || leaf == current);
-	if (leaf == NULL)
-		WT_TRET(__wt_page_release(session, current, 0));
+err:	WT_TRET(__wt_page_release(session, current, 0));
```

Comment:
@keithbostic, I stared at this with @agorrod and couldn't figure out why that test would be needed.

If `leaf == NULL`, then we've found `current` in an ordinary search and it should be released.

If `leaf != NULL`, then we'll jump to `leaf_only` and set `current = NULL`, making `__wt_page_release` a noop.

Is there some other path through here (e.g., via restart) that we need to think through?  That condition has always confused me...